Iox multihome

reicheratwork opened the pull request:
- added ignored locator type field in QoS
- added fixed size topic checking in shared memory checks
- added automatic generation of fixed size topic flag in idlc generated code

eboasson committed:
Multi-homing support + redundant networking

This will (probably) fail if:

* you enable General/RedundantNetworking with more than 3 interfaces per
  reader (on average in the system) (hard-coded multiplier instead of
  counting properly);

* you try to enable TCP or source-specific multicast;

* you include interfaces with link-local IPv6 addresses other
  non-routed/local interfaces among the advertised addresses that aren't
  recognised as such;

* you have interface aliases within the same subnet (I haven't
  considered how it would blow up, but I think it won't like it);

* you have more than 250 reader-interfaces with
  General/RedundantNetworking (hard-coded upper limit on reader "names"
  for making sense of debugging output);

* you advertise multicast addresses that are not listened on by all
  multicast-capable networks, and/or if there's disagreement between
  machines on whether an interface is multicast-capable;

* (possibly something else I can't think of right now)

Some of these bugs disappear if you use only a single interface, but
some details in the behaviour do change compared to the "normal"
version.

But otherwise:

* General/NetworkInterfaceAddress = lo0,en0,en1

  Will make it use these three interfaces while trying to minimize the
  number of packets sent to reach all destinations.  Without specifying
  loopback explicitly it will rely on the kernel routing the packets
  over loopback for local readers (like it has always done).

  Specifying it means it will use it and advertise it to the others, how
  other implementations respond to that, I don't know.

* General/RedundantNetworking = true

  Will make it advertise in a vendor-specific setting that it would like
  to receive data on all advertised interfaces simultaneously, rather
  than on one.

Multi-network support is important, so this is definitely not a wasted
effort; and although it won't be obvious from the current state of
things it'll actually result in a cleaning-up of code.

As far as redundant networking goes, I'm not sure the additional
complexity is worth it compared to the simple one where the transport
simply sends it out twice with a small address transformation in
between.  What this allows (although I haven't made it configurable) is
to make it dependent on the individual readers.  Whether that's valuable
... let's say I have my doubts.

Another downside is that this relies on a discovery protocol extension
and so isn't interoperable with other implementations, whereas always
sending on two wires is interoperable.

Finally, load-balancing would be a useful feature to have as well and
while that is perfectly straightforward when done at the transport level
it is not quite so trivial at this level.

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Split ddsi_locator_t into wire format and internal

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
use gv in locator parsing, store interf in conns

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Take proxypp locators into account for enabling multicast addreses in SEDP

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Update writer address set generation

- Clean up
- Removing arbitrary limits
- Take "reader favours SSM" into account
- All else being equal, prefer locator that reaches most readers

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Allow unicast addresses in network partitions

This commit makes it possible to specify unicast addresses in network
partitions and changes the representation of addresses to a simple list
instead of an addrset.

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Fix networkpartition handling of default partition

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Penalize locators also covering an Iceoryx locator

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Refactoring to clean things up a bit

- Be more consistent in letting transport code handle locator kinds,
  allowing the removal of the locator kind from the common transport type
  definition.

- Replace "ignore" by "enable" in transport.

- Rename "adv_spdp" to "enable_spdp", the first attempt at naming it was
  somewhat ambiguous (advertise or advanced) and it is used in so few
  places that there is no added value in a short name.

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Exclude Iceoryx locators if not enabled for writer

Iceoryx integration is used by a writer iff

- it is included at build time
- it is enabled for the domain in the configuration
- the writer QoS settings and other aspects are are compatible

and so the decision whether or not Iceoryx locators advertised by proxy
readers are applicable needs to be determined per writer.  This adds a
boolean to the computing the writer address set to indicate whether
Iceoryx locators are allowed.

Signed-off-by: Erik Boasson <eb@ilities.com>

eboasson committed:
Fix initialization memory leaks in Iceoryx support

Signed-off-by: Erik Boasson <eb@ilities.com>

reicheratwork committed:
Add ignored locator type in the dds_qos struct

- by setting the locator type (NN_LOCATOR_KIND_XXX) in this field, the discovery
step should ignore locators of this type, through at the moment only
shared memory locators (NN_LOCATOR_KIND_SHEM) are filtered

Signed-off-by: Martijn Reicher <martijn.reicher@adlinktech.com>

reicheratwork committed:
Add filtering and detection of fixed-size topics

- Added automatic generation of the flag DDS_TOPIC_FIXED_SIZE for
topics which do not contain any variable size sequences or strings
in the idlc generated code
- Added filtering such that shared memory is not enabled for topics
which do not have the fixed size flag ddsi_sertype::fixed_size set,
this flag is either initialized by the idlc generated code or set
externally

Signed-off-by: Martijn Reicher <martijn.reicher@adlinktech.com>

eboasson commented on src/core/ddsi/include/dds/ddsi/ddsi_xqos.h:
```diff
@@ -228,11 +228,7 @@ typedef struct dds_type_consistency_enforcement_qospolicy {
-  unsigned char enabled;
-} dds_type_shared_memory_t;
-#endif
+typedef int dds_locator_mask_t;
```
That should be an `uint32_t` — or, more precisely, an unsigned type of sufficient width, with 32 bits usually being a decent compromise. But mind the earlier remarks on using a bitmask to represent the set!

reicheratwork replied:
You are correct on this, using signed entities for bitmasks can give very odd results, and explicitly stating the width is also a plus.
I will correct this.

eboasson commented on src/core/ddsc/src/dds_reader.c:
```diff
@@ -602,16 +602,17 @@ static dds_entity_t dds_create_reader_int (dds_entity_t participant_or_subscribe
-  rqos->shared_memory.enabled = dds_reader_support_shm(&gv->config, rqos);
-  rqos->present |= QP_SHARED_MEMORY;
+  if (!dds_reader_support_shm(&gv->config, rqos))
+    rqos->ignore_locator_type |= NN_LOCATOR_KIND_SHEM;
```
I like the changing the "shared memory" flag to a set of supported locator kinds. There are, however, two (at least 🙂) problems with this:
* The first is minor and a matter of taste: I think I'd find it more elegant if it was a set of enabled locator kinds instead. The downside is of course that the default can no longer be simply 0. It could be that in the end that works out as messier than an ignored set.
* The second is rather major: locator kinds are 32-bit integers with no range limitation, and so a bitset won't do. What I have considered at times is to translate between internal and external locator kinds, because the code only deals with a handful and it generally is attractive to be able to use small integers. Being able to represent them in a small bitmask being one, efficiently looking up the `dds_tran_factory` supporting it being another.

As far as that translation goes, I think it'd be pretty straightforward by now. It already has conversion functions in place (in/near https://github.com/eclipse-cyclonedds/cyclonedds/blob/3108aff3ee8c8b4cdb418441d1fe0925de3ef273/src/core/ddsi/src/ddsi_plist.c#L376), and no locators ever go in/out without going through these functions. So implementing this translation should be no more than a limited bit of refactoring.

reicheratwork replied:
I was thinking of using "allowed" locator kinds first, but since this caused issues with initializing the default, I used the inverse, so a "disallowed", mask first.

I will look into the conversion functions, since this will alow this mechanism to become more generic.

eboasson commented on src/core/ddsi/src/ddsi_plist.c:
```diff
@@ -1761,11 +1761,9 @@ static const struct piddesc piddesc_eclipse[] = {
-#endif
+    { PID_PAD, PDF_QOS, QP_LOCATOR_MASK, "CYCLONE_LOCATOR_MASK",
+    offsetof(struct ddsi_plist, qos.ignore_locator_type), membersize(struct ddsi_plist, qos.ignore_locator_type),
+    {.desc = { Xi, XSTOP } }, 0 },
```
Changing the bitmask type to a `uint32_t` means this should be `Xu`.

reicheratwork replied:
Yep.

eboasson commented on src/core/ddsi/src/ddsi_plist.c:
```diff
@@ -3280,6 +3278,9 @@ void ddsi_xqos_init_default_reader (dds_qos_t *xqos)
   xqos->type_consistency.force_type_validation = false;
+
+  xqos->present |= QP_LOCATOR_MASK;
+  xqos->ignore_locator_type = 0;
```
I agree with the decision to add this here; but then in `dds_create_reader` and `dds_create_writer`, it must be present already and there is no need to add `QP_LOCATOR_MASK` another time.

reicheratwork replied:
You are correct in this.

eboasson commented on src/core/ddsi/src/ddsi_wraddrset.c:
```diff
@@ -347,14 +347,17 @@ static readercount_cost_t calc_locator_cost (const struct cover *c, int lidx, bo
   if (rdidx == c->nreaders)
     goto no_readers;
 
+#ifdef DDS_HAS_SHM
```
Is there really a need to do `#ifdef DDS_HAS_SHM` here? I had tried to avoid those as much as I could, so it is a bit of a shame if it needs to be re-added. It perhaps would make more sense to define `NN_LOCATOR_KIND_SHEM` unconditionally in `q_protocol.h`.

reicheratwork replied:
At the moment I put this inside the ifdefs because NN_LOCATOR_KIND_SHEM is only defined when this compile flag is set.

I was more thinking in the direction of a more generic checking function which compares the locator kind together with the mask and makes a decision there.

eboasson replied:
Good point, it would probably be vastly more sensible to store the locator type directly in an accessible place (we have `struct locset` after all) and simply check whether the kind is included/not excluded.

Or, perhaps even wiser, bite the bullet and check the locator kinds in all places where it deals with locators as stored in the readers' address sets — that is, `wras_flatten_locs_prealloc` and expanding `copy_addrset_into_addrset` in `wras_collect_all_locs` and filtering right there (or modify `wras_calc_locators`).

eboasson commented on src/core/ddsi/src/ddsi_wraddrset.c:
```diff
@@ -784,7 +787,7 @@ struct addrset *compute_writer_addrset (const struct writer *wr, bool enable_shm
   else
   {
-    struct costmap *wm = wras_calc_costmap (covered, prefer_multicast, enable_shm);
+    struct costmap *wm = wras_calc_costmap (covered, prefer_multicast, (wr->xqos->present & QP_LOCATOR_MASK) ? wr->xqos->ignore_locator_type : 0);
```
With the locator mask always present in the QoS, there's no need for the conditional anymore. (I personally would probably add an assert: those things verify invariants/assumptions and usually reduce the amount of state I have to track in my mind, instead of enlarging it like this does.)

reicheratwork replied:
Maybe the assert can be put inside the validate_reader_qos function?
Then that function does what it says on the tin? :)

eboasson replied:
🙂

But no, not really. `validate_reader_qos` is about validating input, this is about verifying that some invariant (in this case, reader/writer->xqos actually has the locator mask initialized) holds. If you check every invariant everywhere, it becomes really easy to write flawless software, the only thing is that you never get it done 🙂

In this case, I think it would be useful at point of use, because things will go horrible wrong in hard to detect ways if it is uninitialized, and if such a check exists in a few key places any screw ups will get caught quickly.

reicheratwork replied:
I agree with your point about not overchecking things, but I personally do not like using asserts to validate input, while they are fine in debug builds, they are a little too unreliable for release builds in my eyes.

eboasson replied:
For general input validation, they are pretty bad, no doubt about that. For checking assumptions and invariants, they are (unfortunately) the best we have in C, because a real run-time check means you get a system that behaves differently from what you expect. In the end that really makes it harder to reason about which states are valid, and which aren't. In my experience, in the long run, that means you give up and you simply *assume* that certain states are valid because it is tested for somewhere and you end up propagating the assumption to other places until eventually no-one remembers that, actually, that state is impossible. (That was really a painful lesson to learn ...)

I can live with not checking at all (not even assert — but that assertion is essentially free, so why not?), but I do have a problem with handling the `!(present & QP_LOCATOR_MASK)` case (here), when it cannot be because it is included in the default set.

If that makes any sense 🙂

eboasson commented on src/core/ddsi/src/q_ddsi_discovery.c:
```diff
@@ -1161,8 +1161,8 @@ static int sedp_write_endpoint_impl
 #ifdef DDS_HAS_SHM
-    if ((xqos->present & QP_SHARED_MEMORY) == QP_SHARED_MEMORY &&
-      xqos->shared_memory.enabled)
+    if (xqos->present & QP_LOCATOR_MASK &&
```
Same here, no need for the `xqos->present & QP_LOCATOR_MASK` anymore

reicheratwork replied:
Indeed.

eboasson commented on src/core/ddsc/src/dds_topic.c:
```diff
@@ -596,6 +596,7 @@ dds_entity_t dds_create_topic (dds_entity_t participant, const dds_topic_descrip
 #ifdef DDS_HAS_SHM
   st->c.iox_size = desc->m_size;
 #endif
+  st->c.fixed_size = (st->c.fixed_size || (desc->m_flagset & DDS_TOPIC_FIXED_SIZE)) ? 1u : 0u;
```
Instead of `ddsi_sertype_init` + setting the `fixed_size` flag afterward, it'd probably make more sense to use `ddsi_sertype_init_flags`. The `iox_size` bit still needs to be integrated as well, but let's not add more "fix ups" after `ddsi_sertype_init` than necessary!

eboasson commented on src/tools/idlc/src/descriptor.c:
```diff
@@ -1287,7 +1287,7 @@ static int print_keys(FILE *fp, struct descriptor *descriptor, bool keylist)
 {
   const char *fmt;
-  const char *vec[3] = { NULL, NULL, NULL };
+  const char *vec[4] = { NULL, NULL, NULL, NULL };
```
I'd do `const char *vec[4] = { NULL };` because that guarantees all remaining entries in `vec` are initialized to a null pointer as well.

reicheratwork replied:
Yep.

eboasson commented on src/tools/idlc/src/descriptor.c:
```diff
@@ -1297,6 +1297,22 @@ static int print_flags(FILE *fp, struct descriptor *descriptor)
+    if (((i.data.opcode.code>>16)&0xFF) == DDS_OP_VAL_STR ||
+      ((i.data.opcode.code >> 16) & 0xFF) == DDS_OP_VAL_BST ||
+      ((i.data.opcode.code >> 16) & 0xFF) == DDS_OP_VAL_SEQ)
+      fixed_size = false;
```
A few points:
* Shouldn't this also recurse into structs & unions?
* I also suspect there are functions or macros to extract the opcode (there are definitions in `ddsi_cdrstream.h`, but I'm guessing the IDL compiler already has some)
* The `flags` field was inherited, but some things were missing in the old IDL compiler and some fields were computed incorrectly. Instead of fixing that old piece of junk, I instead started computing (some of) them while constructing the `ddsi_sertype_default`, e.g., `dds_stream_check_optimize` & `dds_stream_countops` in `ddsi_cdrstream.c`. Doing it twice is a bit silly, but a run-time verification really is required if the types can ever come from an untrusted source (which is definitely possible nowadays with the type lookup capability). So I would like to consider doing it in `ddsi_cdrstream.c` instead.

reicheratwork replied:
I thought that the current iteration of the opcode generation fully unrolls all structures, so that by walking over the complete list of opcodes of a structure you would be able to see "into" all present structs and unions as well, but I will examine this further.
I will look into ddsi_cdrstream.c and see what kind of inspiration I am able to glean from there.

reicheratwork replied:
As far as I can see, all structures are fully unrolled in the opcodes

eboasson reviewed (commented):
Of course there are some pitfalls and traps that manage to catch the unwary. I never meant to put those on the paths you (correctly) chose to tread, but I am afraid I did put them there anyway. It is really just about the gory details, I definitely like the direction you travelled in!

reicheratwork committed:
Fix for iceoryx locator filtering and adding process id to locator

- The process id is NOT added to the iceoryx locator, since this
would cause cyclone to see different processes' iceoryx services
on the same node as different instances, which they are not
- The address of the iceoryx locator is checked, and ignored if
this is not the local iceoryx instance, since this would cause cyclone
to not be able to differentiate between local and remote iceoryx
services

Signed-off-by: Martijn Reicher <martijn.reicher@adlinktech.com>

eboasson commented:
@reicheratwork I have merged it into my `iox-multihome` branch by cherry-picking it because of various other things I had (force-)pushed.
